### PR TITLE
MOTECH-2163 Add metrics module.

### PR DESCRIPTION
**Pull request by justin-hayes**
Integrates the Dropwizard metrics library with the core platform,
exposes an API to register metrics with the system,
and allows implementers to stream metrics to Graphite.

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by pgesek** [comment]
test this please

File: modules/metrics/pom.xml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            <groupId>io.dropwizard.metrics</groupId>
+            <artifactId>metrics-graphite</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-webmvc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-test-mvc</artifactId>
```
Comment (sebbrudzinski) [inline modules/metrics/pom.xml]: Should this be in the test scope?

File: modules/metrics/pom.xml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                            org.springframework.aop.framework,
+                            *
+                        </Import-Package>
+                        <Export-Package>
+                            org.motechproject.metrics.api;version=${project.version},
+                            org.motechproject.metrics.builder;version=${project.version},
+                            org.motechproject.metrics.service;version=${project.version}
+                        </Export-Package>
+                        <Bundle-DocURL>
+                            http://docs.motechproject.org
```
Comment (sebbrudzinski) [inline modules/metrics/pom.xml]: Do not link to the main docs page if this module does not have any documentation yet

File: modules/metrics/src/main/java/org/motechproject/metrics/event/MotechEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import org.motechproject.metrics.api.Counter;
+import org.motechproject.metrics.service.MetricRegistryService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+/**
+ * Handles the modification of metrics through the event bus or Tasks interface.
+ */
+@Service
+public class MotechEventHandler {
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/event/MotechEventHandler.java]: Rename this class to indicate this is a handler for the metrics module, for example MetricsEventHandler

File: modules/metrics/src/main/java/org/motechproject/metrics/exception/MetricNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.motechproject.metrics.exception;
+
+/**
+ * A custom exception thrown when a metric does not exist by name in the metric registry, and the default behavior
+ * of creating a new metric of the appropriate type and name is not an appropriate solution.
+ */
+public class MetricNotFoundException extends RuntimeException {
+
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/exception/MetricNotFoundException.java]: Lacks serialVersionUID

File: modules/metrics/src/main/java/org/motechproject/metrics/model/GaugeAdapter.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package org.motechproject.metrics.model;
+
+/**
+ * An implementation of a Gauge.
+ *
+ * @param <T> the return type when getting the gauge's value
+ */
+public class GaugeAdapter<T> implements org.motechproject.metrics.api.Gauge<T> {
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/model/GaugeAdapter.java]: You should import one of the Gauge classes

File: modules/metrics/src/main/java/org/motechproject/metrics/service/MetricRegistryService.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+     * Register a ratio gauge.
+     *
+     * @param name the name of the gauge
+     * @param numerator a function returning a number represents the value of the numerator
+     * @param denominator a function returning a number that represents the value of the denominator
+     * @param <T> a type of number
+     * @return the registered gauge
+     */
+    <T extends Number> Gauge<Double> registerRatioGauge(final String name, Supplier<T> numerator, Supplier<T> denominator);
+    boolean isRegistered(String name);
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/MetricRegistryService.java]: Needs javadoc and a line of space from the method signature above

File: modules/metrics/src/main/java/org/motechproject/metrics/service/impl/HealthCheckRegistryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    public HealthCheckRegistryServiceImpl(HealthCheckRegistry healthCheckRegistry) {
+        this.healthCheckRegistry = healthCheckRegistry;
+    }
+
+    @Override
+    public void register(String name, HealthCheck healthCheck) {
+        healthCheckRegistry.register(name, new com.codahale.metrics.health.HealthCheck() {
+            @Override
+            @SuppressWarnings("PMD.SignatureDeclareThrowsException")
+            protected Result check() throws Exception {
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/impl/HealthCheckRegistryServiceImpl.java]: Why exactly is it marked as "throws Exception"?

File: modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        com.codahale.metrics.RatioGauge theGauge;
+        try {
+            theGauge = metricRegistry.register(name, new RatioGauge() {
+                @Override
+                protected Ratio getRatio() {
+                    return Ratio.of(numerator.get().doubleValue(), denominator.get().doubleValue());
+                }
+            });
+        } catch (IllegalArgumentException ex) {
+            throw ex;
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java]: Doesn't make sense to catch the exception only to throw the same exception again. Did you mean to wrap it into custom exception?

File: modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        com.codahale.metrics.Gauge<T> theGauge;
+        try {
+            theGauge = metricRegistry.register(name, new com.codahale.metrics.Gauge<T>() {
+                @Override
+                public T getValue() {
+                    return gauge.getValue();
+                }
+            });
+        } catch (IllegalArgumentException ex) {
+            throw ex;
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java]: Doesn't make sense to catch the exception only to throw the same exception again. Did you mean to wrap it into custom exception?

File: modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    private <T extends Metric> T getOrAdd(String name, MetricBuilder<T> builder) {
+        Metric metric = metrics.get(name);
+
+        if (metric == null) {
+            try {
+                T built = builder.createMetric(name, metricRegistry, metricsConfigFacade);
+                metrics.put(name, built);
+                return built;
+            } catch (IllegalArgumentException ex) {
+                throw ex;
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/impl/MetricRegistryServiceImpl.java]: Doesn't make sense to catch the exception only to throw the same exception again. Did you mean to wrap it into custom exception?

File: modules/metrics/src/main/java/org/motechproject/metrics/web/HealthCheckRegistryController.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Initiates health checks from the user interface and returns the result.
+ */
+@Controller
+@PreAuthorize(HAS_VIEW_METRICS_ROLE)
+public class HealthCheckRegistryController {
+    private HealthCheckRegistry healthCheckRegistry;
+
+    public HealthCheckRegistryController() {}
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/web/HealthCheckRegistryController.java]: Remove

File: modules/metrics/src/main/java/org/motechproject/metrics/web/MetricsConfigController.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Sends and receives configuration to and from the user interface.
+ */
+@Controller
+@PreAuthorize(HAS_MANAGE_METRICS_ROLE)
+public class MetricsConfigController {
+    private MetricsConfigFacade metricsConfigFacade;
+    private MetricRegistryInitializer metricRegistryInitializer;
+
+    public MetricsConfigController() {}
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/web/MetricsConfigController.java]: Remove

File: modules/metrics/src/main/java/org/motechproject/metrics/web/MetricsController.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * Sends details about the currently registered metrics to the user interface.
+ */
+@Controller
+public class MetricsController {
+    private MetricRegistry metricRegistry;
+    private MetricRegistryService metricRegistryService;
+    private MetricDtoToSupplierConverter converter;
+
+    public MetricsController() {}
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/web/MetricsController.java]: Remove

File: modules/metrics/src/main/java/org/motechproject/metrics/web/dto/MetricsDto.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * A list of names of all metrics of that type.
+     */
+    private SortedSet<String> names;
+
+    /**
+     * The set of possible values
+     */
+    private RatioGaugeValue[] values;
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/web/dto/MetricsDto.java]: Can we use a Set here?

File: modules/metrics/src/main/resources/task-channel.json [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    {
+      "subject" : "metrics_histogram_update",
+      "displayName" : "metrics.task.update.histogram",
+      "actionParameters" : [
+        {
+          "key" : "metric_name",
+          "displayName" : "metrics.task.histogram.name"
+        },
+        {
+          "key" : "metric_name",
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/resources/task-channel.json]: seems like it should be metric_value

File: modules/metrics/src/main/resources/webapp/partials/metrics.html [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+                <div class="form-group">
+                    <label class="col-md-2 col-sm-3 control-label" for="ratioGaugeNumeratorType">Numerator type</label>
+                    <div class="form-inline col-md-10 col-sm-9">
+                        <select class="form-control input-auto"
+                                required
+                               ng-model="ratioGauge.numerator.type"
+                               ng-change="setSelectedNumerator()"
+                               ng-options="metric.type as metric.type for metric in metrics"
+                               id="ratioGaugeNumeratorType">
+                            <option value="">---Select type---</option>
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/resources/webapp/partials/metrics.html]: I'd remove the --- and move it to the message.properties

File: modules/metrics/src/test/java/org.motechproject.metrics/it/HealthCheckRegistryServiceBundleIT.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+@RunWith(PaxExam.class)
+@ExamReactorStrategy(PerSuite.class)
+@ExamFactory(MotechNativeTestContainerFactory.class)
+public class HealthCheckRegistryServiceBundleIT {
+    @Inject
+    private HealthCheckRegistryService healthCheckRegistryService;
+
+    @Test
+    public void shouldRegisterHealthCheckRegistryServiceInstance() {
+        assertNotNull(healthCheckRegistryService);
```
Comment (sebbrudzinski) [inline modules/metrics/src/test/java/org.motechproject.metrics/it/HealthCheckRegistryServiceBundleIT.java]: Is it worth to set up a whole framework, just to check if PaxIT injects a service? This should either be a real test or it should be removed.

File: modules/metrics/src/test/java/org.motechproject.metrics/ut/config/MetricsConfigFacadeTest.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
+package org.motechproject.metrics.ut.config;
```
Comment (sebbrudzinski) [inline modules/metrics/src/test/java/org.motechproject.metrics/ut/config/MetricsConfigFacadeTest.java]: It is kinda an unwritten rule, but we keep the unit tests in the same package as the tested class.

File: modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/TaskTriggerHandler.java [hunk 19]
```diff
@@ -176,6 +182,7 @@ private boolean isTaskRetryAlreadyScheduled(Map<String, Object> eventParams) {
     }
 
     private void handleTask(Task task, Map<String, Object> parameters) {
+        metricRegistryService.meter(TASKS_TRIGGER_HANDLER_METER);
```
Comment (sebbrudzinski) [inline modules/tasks/tasks/src/main/java/org/motechproject/tasks/service/TaskTriggerHandler.java]: Is it some leftover from testing or are we planning to introuce the metrics module here? Is there some requirement for this? The Task module already exposes the task activity tracking service.

File: platform/server/pom.xml [hunk 20]
```diff
@@ -114,6 +114,12 @@
             <scope>provided</scope>
         </dependency>
         <dependency>
+            <groupId>org.motechproject</groupId>
+            <artifactId>motech-metrics</artifactId>
+            <version>${project.version}</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (sebbrudzinski) [inline platform/server/pom.xml]: Any reason for this module to be a platform bundle?

File: modules/metrics/pom.xml [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+                        </Bundle-DocURL>
+                    </instructions>
+                </configuration>
+            </plugin>
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-dependency-plugin</artifactId>
+            </plugin>
+        </plugins>
```
Comment (sebbrudzinski) [inline modules/metrics/pom.xml]: Lacks maven-resources-plugin, that copies the module to the .motech/bundles directory. Currently it only gets copied due to the fact, that the module is marked as platform(!) bundle.

Comment (sebbrudzinski) [inline modules/metrics/pom.xml]: All our modules introduce the maven-clean-plugin to clear the settings directory during build. We should probably stick to that here as well.

Comment (sebbrudzinski) [inline modules/metrics/pom.xml]: Lacks jslint-maven-plugin that runs the checks on the javascript files in our modules

**Comment by sebbrudzinski** [comment]
There are some conflicts that must be resolved.

File: modules/metrics/src/main/resources/securityRules.json [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    "active": true,
+    "pattern": "/**/metrics/api/**",
+    "supportedSchemes": [
+      "BASIC"
+    ],
+    "protocol": "HTTP",
+    "priority": 2,
+    "rest": true,
+    "origin": "SYSTEM_MODULE_METRICS",
+    "version": "1.0-SNAPSHOT",
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/resources/securityRules.json]: Just mark it as 1.0

File: modules/metrics/src/main/java/org/motechproject/metrics/service/MetricRegistryService.java [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+
+import org.motechproject.metrics.api.Counter;
+import org.motechproject.metrics.api.Gauge;
+import org.motechproject.metrics.api.Histogram;
+import org.motechproject.metrics.api.Meter;
+import org.motechproject.metrics.api.Timer;
+
+import java.util.function.Supplier;
+
+public interface MetricRegistryService {
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/MetricRegistryService.java]: Lacks javadoc

File: modules/metrics/src/main/java/org/motechproject/metrics/service/HealthCheckRegistryService.java [hunk 24]
```diff
@@ -0,0 +1,29 @@
+package org.motechproject.metrics.service;
+
+import org.motechproject.metrics.api.HealthCheck;
+
+import java.util.SortedSet;
+
+public interface HealthCheckRegistryService {
```
Comment (sebbrudzinski) [inline modules/metrics/src/main/java/org/motechproject/metrics/service/HealthCheckRegistryService.java]: Lacks javadoc

**Comment by justin-hayes** [comment]
Submitted a new pull request to reflect moving this module from the core platform to the modules repository: https://github.com/motech/modules/pull/192

**Pull request closed without merging**